### PR TITLE
Implement LLM Registry

### PR DESCRIPTION
**Pull request by MarcCote**
The idea is with a global LLM Registry, it will be easier to get the total_cost (in number of tokens) used by the different LLMs during training and at test.

### CODE REVIEW

File: dln/operator.py [hunk 1]
```diff
@@ -361,21 +378,66 @@ def instantiate_tokenizer(model_name: str):
...
+        self.models[model_name] = llm
         return llm
 
     @property
     def total_cost(self):
-        return sum([llm.total_cost for llm in self.models])
+        return sum([llm.total_cost for llm in self.models.values()])
+
+    @classmethod
+    def instantiate_llm(cls, model_type: str, **generation_options) -> LLM:
```
Comment (MarcCote) [inline dln/operator.py:409]: What is the purpose of having both register and instantiate_llm?

File: dln/operator.py [hunk 2]
```diff
@@ -361,21 +378,66 @@ def instantiate_tokenizer(model_name: str):
...
+    def register(self, model_name: str, model_type: str = None, **generation_options) -> LLM:
+        """ Register a single model to the registry.
+        Args:
+            model_name: how you refer to the model, for example: gpt-3.
+            model_type: the api model name, for example: text-davinci-003. If not provided, use model_name as default.
+            **generation_options: generation options, for example: api_key, api_base, api_type, api_version, max_tokens, temperature, etc.
+        Returns:
+            the instantiated model
+        """
+        if model_type is None:
```
Comment (MarcCote) [inline dln/operator.py:398]: What happens if that model name was already registered?

File: dln/vi/model.py [hunk 3]
```diff
@@ -709,7 +710,7 @@ def forward(self, x, y=None, infos=None, temperature=0.0, cost_only=False):
                 x_stripped, temperature=temperature, max_tokens=self.p1_max_tokens
             )
 
-            self.cost += self.encoder_l2.forward_evaluate.compute_cost(x_stripped)
+            self.cost += self.forward_evaluate.compute_cost(x_stripped)
```
Comment (MarcCote) [inline dln/vi/model.py:713]: Is that needed? I thought the purpose of the registry was to avoid having to track the cost like this.

**Review by MarcCote — COMMENTED** [review]
Got some comments for you @matheper

File: dln/operator.py (hunk 2 above)
Comment (matheper) [reply dln/operator.py:398]: At the moment it overrides the previous model, but we can raise an exception instead. Which one makes more sense for you?

Comment (MarcCote) [reply dln/operator.py:398]: It should have a unique UUID. Do you see any issues with having multiple instances of a same model?

File: dln/operator.py (hunk 1 above)
Comment (matheper) [reply dln/operator.py:409]: I can remove `instantiate_llm` if it is confusing. This function is just a syntactic sugar, so you don't need to instantiate the registry to get an LLM.

Comment (MarcCote) [reply dln/operator.py:409]: We could make one singleton instance of the Registry.

File: dln/vi/model.py (hunk 3 above)
Comment (matheper) [reply dln/vi/model.py]: Yes, using the registry this is not required.

File: dln/operator.py (hunk 2 above)
Comment (matheper) [reply dln/operator.py:384]: I don't see any issue with having multiple instances of the same model. We register the same model using different default generation options (temperature) for backward and forward models, for example.

Comment (matheper) [reply dln/operator.py:384]: I would prefer the user to specify a name for every llm registered, and the register to raise an exception if the `model_name` already exists. The use case I have in mind (from llama2_scripts branch, and that is partially implemented here) is to register many models from a config file and refer to them by name everywhere else. What do you think?

Comment (MarcCote) [reply dln/operator.py:380]: That makes sense. I agree it is better to let the user have entire control over the model name.

Comment (matheper) [reply dln/operator.py:384]: Sounds good. Register now raises an exception if model name already exists.

File: dln/operator.py (hunk 1 above)
Comment (matheper) [reply dln/operator.py]: I would be fine if we didn't implement the singleton for now. What do you think?

**Review by matheper — APPROVED** [review]

**Review by MarcCote — COMMENTED** [review]
Thanks for beefing up my PR @matheper  :D [Approved].

**Pull request merged**